source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc

```diff
@@ -413,6 +423,49 @@ DnsCacheImpl::PrimaryHostInfo::~PrimaryHostInfo() {
   parent_.stats_.num_hosts_.dec();
 }
 
+void DnsCacheImpl::addCacheEntry(const std::string& host,
+                                 const Network::Address::InstanceConstSharedPtr& address) {
+  if (!key_value_store_) {
+    return;
+  }
+  // TODO(alyssawilk) cache data should include TTL, or some other indicator.
+  const std::string value = absl::StrCat(address->asString());
+  key_value_store_->addOrUpdate(host, value);
+}
+
+void DnsCacheImpl::removeCacheEntry(const std::string& host) {
+  if (!key_value_store_) {
+    return;
+  }
+  key_value_store_->remove(host);
+}
+
+void DnsCacheImpl::loadCacheEntries(
+    const envoy::extensions::common::dynamic_forward_proxy::v3::DnsCacheConfig& config) {
+  if (!config.has_key_value_config()) {
+    return;
+  }
+  auto& factory =
+      Config::Utility::getAndCheckFactory<KeyValueStoreFactory>(config.key_value_config().config());
+  key_value_store_ = factory.createStore(config.key_value_config(), validation_visitor_,
+                                         main_thread_dispatcher_, file_system_);
+  KeyValueStore::ConstIterateCb load = [this](const std::string& key, const std::string& value) {
+    auto address = Network::Utility::parseInternetAddressAndPortNoThrow(value);
+    if (address == nullptr) {
+      return KeyValueStore::Iterate::Break;
+    }
+    stats_.cache_load_.inc();
+    std::list<Network::DnsResponse> response;
+    // TODO(alyssawilk) change finishResolve to actually use the TTL rather than
+    // putting 0 here, return the remaining TTL or indicate the result is stale.
+    response.emplace_back(Network::DnsResponse(address, std::chrono::seconds(0) /* ttl */));
+    startCacheLoad(key, address->ip()->port());
```

Comment:
I'm not that familiar with this code, but it looks like this will kick off a DNS query? And then we immediately resolve it? 

If so I wonder if this will cause unforeseen consequences, like having entries being resolved twice (once from cache and once from DNS) and confusing stats (e.g. stat increment for DNS resolution and cache load for same entry)

Thoughts?

Reply:
it works today (see integration test) but I agree it's not great, hence the TODO :-)

What I want to have happen is be able to differentiate stale cached results from fresh.  I do want to immediately kick off a resolve, but also have stale data in there, config for how long we wait for fresh data before we use stale data.  I plan to do that in the follow up

Reply:
For my understanding, would a DNS resolution then overwrite this value once it arrives?

Reply:
yeah, fresh resolve always overwrites older data (if it's different)
in this case it'd be the same so not trigger any changes.